fish-rust/src/builtins/math.rs

```diff
@@ -0,0 +1,319 @@
+}
+
+/// We read from stdin if we are the second or later process in a pipeline.
+fn use_args_from_stdin(streams: &io_streams_t) -> bool {
+    streams.stdin_is_directly_redirected()
+}
+
+/// Get the arguments from stdin.
+fn get_arg_from_stdin(streams: &io_streams_t) -> Option<WString> {
+    let mut s = WString::new();
+    loop {
+        let mut buf = [0];
+        let c = match read_blocked(streams.stdin_fd().unwrap().as_raw_fd(), &mut buf) {
+            1 => char::from(buf[0]),
+            0 => {
+                // EOF
+                if s.is_empty() {
+                    return None;
+                } else {
+                    break;
+                }
+            }
+            n if n < 0 => {
+                // error
+                perror("read");
+                return None;
+            }
+            n => panic!("Unexpected return value from read_blocked(): {n}"),
+        };
+
+        if c == '\n' {
+            // we're done
+            break;
+        }
+
+        s.push(c);
+    }
+
+    Some(s)
+}
+
+/// Get the arguments from argv or stdin based on the execution context. This mimics how builtin
+/// `string` does it.
+fn get_arg<'args>(
+    argidx: &mut usize,
+    args: &'args [&'args wstr],
+    streams: &io_streams_t,
+) -> Option<Cow<'args, wstr>> {
+    if use_args_from_stdin(streams) {
+        assert!(
+            streams.stdin_fd().is_some(),
+            "stdin should not be closed since it is directly redirected"
+        );
+
+        get_arg_from_stdin(streams).map(Cow::Owned)
+    } else {
+        let ret = args.get(*argidx).copied().map(Cow::Borrowed);
+        *argidx += 1;
+        ret
+    }
+}
+
+/// Return a formatted version of the value `v` respecting the given `opts`.
+fn format_double(mut v: f64, opts: &Options) -> WString {
+    if opts.base == 16 {
+        v = v.trunc();
+        let mneg = if v.is_sign_negative() { "-" } else { "" };
+        return sprintf!(L!("%s0x%lx"), mneg, v.abs() as u64);
+    } else if opts.base == 8 {
+        v = v.trunc();
+        if v == 0.0 {
+            // not 00
+            return WString::from_str("0");
+        }
+        let mneg = if v.is_sign_negative() { "-" } else { "" };
+        return sprintf!(L!("%s0%lo"), mneg, v.abs() as u64);
+    }
+
+    // As a special-case, a scale of 0 means to truncate to an integer
+    // instead of rounding.
+    if opts.scale == 0 {
+        v = v.trunc();
+        return sprintf!(L!("%.*f"), opts.scale, v);
+    }
+
+    let mut ret = sprintf!(L!("%.*f"), opts.scale, v);
+    // If we contain a decimal separator, trim trailing zeros after it, and then the separator
+    // itself if there's nothing after it. Detect a decimal separator as a non-digit.
+    if ret.chars().any(|c| !c.is_ascii_digit()) {
+        let trailing_zeroes = ret.chars().rev().take_while(|&c| c == '0').count();
+        let mut to_keep = ret.len() - trailing_zeroes;
+        if ret.as_char_slice()[to_keep - 1] == '.' {
```

Comment:
This code used to handle alternative decimal separators. I don't think they are still supported by `math` so this should be fine

Reply:
Specifically, we currently set our internal LC_NUMERIC to C so *all* our float reading only ever uses `.`, and for math specifically alternative separators were never supported and the popular `,` *cannot* be supported because that's the argument separator (german Excel switches to `;` so it can support `,` decimal separators).

Reply:
Note that this code looks at the *output* of `sprintf!()`, which currently always uses `C_LOCALE`, but could use something else as well - I guess that does need special handling then?

Reply:
Depends - does https://github.com/fish-shell/fish-shell/blob/971d257e67bc6482f5a3dbeba5fd70dcb0265029/fish-rust/src/locale.rs#L119-L121 affect the decimal separator or not?

If it doesn't, we need some other way to change the decimal separator here to always always always be `.` specifically.

Reply:
That line doesn't seem to do anything, it changes the locale after `lconv_to_locale()` has already populated `result` with the locale's decimal separator. Is that a bug or am I missing something?

Reply:
I've not kept a close eye on the locale port.

What it is *supposed* to do is to set the process' locale so that everything but the "LC_NUMERIC" category is taken from the environment, while LC_NUMERIC is set to C.

This is, of course, assuming c-like semantics where the global locale affects all sorts of functions - so something like strtod/wcstod would automatically be affected after the `setlocale`.

If that isn't the case, here, this needs to be changed. (and if `libc::setlocale` is ineffective in that way, it would be cool if that were documented)

We want to have a decimal separator of `.` almost everywhere.

Reply:
Oh, I see, it does affect the C/C++ standard library functions that look directly at the locale. Anyway, `math` has always used C numeric locale for its output, and it still does now due to `sprintf!()` always using C locale. I think this is fine as is then.

Reply:
See also the [locale-numeric test](https://github.com/fish-shell/fish-shell/blob/master/tests/checks/locale-numeric.fish).